src/Routing/Middleware/RedirectMiddleware.php

```diff
@@ -0,0 +1,51 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.1.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Routing\Middleware;
+
+use Cake\Routing\Exception\RedirectException;
+use Laminas\Diactoros\Response\RedirectResponse;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+/**
+ * Ensures any RedirectException thrown is redirected.
+ */
+class RedirectMiddleware implements MiddlewareInterface
+{
+    /**
+     * Ensures any RedirectException thrown is redirected.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request.
+     * @param \Psr\Http\Server\RequestHandlerInterface $handler The request handler.
+     * @return \Psr\Http\Message\ResponseInterface A response.
+     */
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
```

Comment:
Do we need a Middleware for this? Handling `RedirectException` seems like behavior we could build into `ControllerFactory::invoke()` or into Http/Server. If we handled it in Http/Server we would need to update the ErrorHandlerMiddleware to not swallow those errors which is a bit clunky and not idea in my opinion.

Reply:
I am also fine with that :)
This was just my first idea.

Reply:
Throwing the exception from a component or controller class is another opt-in mechanism :smile:

One drawback to handing `RedirectException` in the `ControllerFactory` is that RedirectExceptions raised in middleware won't work the same.  Perhaps the `ErrorHandlerMiddleware` is the best place for these exceptions to be handled. It already handles lots of other kinds of exceptions, and this one has a pretty straightforward action to take.

Should we also expand the `RedirectException` to allow headers to be added?

Reply:
I am fine if someone draws up a better alternative PR.